src/Symfony/Component/HttpKernel/HttpCache/Psr6Store.php

```diff
@@ -0,0 +1,242 @@
+     */
+    public function lock(Request $request)
+    {
+        $lockKey = $this->getLockKey($request);
+
+        if (isset($this->locks[$lockKey])) {
+            return true;
+        }
+
+        $item = $this->cachePool->getItem($lockKey);
+
+        if ($item->isHit()) {
+            return false;
+        }
+
+        $this->cachePool->save($item);
+
+        $this->locks[$lockKey] = true;
+
+        return true;
+    }
+
+    /**
+     * Releases the lock for the given Request.
+     *
+     * @param Request $request A Request instance
+     *
+     * @return bool False if the lock file does not exist or cannot be unlocked, true otherwise
+     */
+    public function unlock(Request $request)
+    {
+        $lockKey = $this->getLockKey($request);
+
+        if (!isset($this->locks[$lockKey])) {
+            return false;
+        }
+
+        $this->cachePool->deleteItem($lockKey);
+
+        unset($this->locks[$lockKey]);
+
+        return true;
+    }
+
+    /**
+     * Returns whether or not a lock exists.
+     *
+     * @param Request $request A Request instance
+     *
+     * @return bool true if lock exists, false otherwise
+     */
+    public function isLocked(Request $request)
+    {
+        $lockKey = $this->getLockKey($request);
+
+        if (isset($this->locks[$lockKey])) {
+            return true;
+        }
+
+        return $this->cachePool->hasItem($this->getLockKey($request));
+    }
+
+    /**
+     * Purges data for the given URL.
+     *
+     * @param string $url A URL
+     *
+     * @return bool true if the URL exists and has been purged, false otherwise
+     */
+    public function purge($url)
+    {
+        // TODO: Implement purge() method.
+    }
+
+    /**
+     * Cleanups storage.
+     */
+    public function cleanup()
+    {
+        $this->cachePool->deleteItems(array_keys($this->locks));
+        $this->locks = array();
+    }
+
+    /**
+     * @param Request $request
+     *
+     * @return string
+     */
+    private function getCacheKey(Request $request)
+    {
+        return 'md'.hash('sha256', $request->getUri());
+    }
+
+    /**
+     * @param Request $request
+     *
+     * @return string
+     */
+    private function getLockKey(Request $request)
+    {
+        return $this->getCacheKey($request).'.lock';
+    }
+
+    /**
+     * @param Response $response
+     *
+     * @return string
+     */
+    private function generateContentDigest(Response $response)
+    {
+        return 'en'.hash('sha256', $response->getContent());
+    }
+
+    /**
+     * @param string $key
+     * @param string $data
+     *
+     * @return bool
+     */
+    private function save($key, $data)
+    {
+        return $this->cachePool->save($this->createCacheItem($key, $data));
+    }
+
+    /**
+     * @param string $key
+     * @param mixed  $value
+     * @param bool   $isHit
+     *
+     * @return CacheItem
+     */
+    private function createCacheItem($key, $value, $isHit = false)
```

Comment:
This is kind of feedback i wanted to get (ignore comments etc, i left them there for my reference until I'm done and replace them with inheritdoc) ;)

> You must get the item from the pool first 

I did it first, perhaps I tried to optimise too early and replaced a call to pool with this. I'll revert back. Once it's all working we can run tests and see if it needs improving.